### ISSUE TITLE
Add support for last -F

### ISSUE THREAD

**Issue by kellyjonbrazil** [kellyjonbrazil/jc#96]
Support planned for v1.14.1. New `last` parser is here:

https://github.com/kellyjonbrazil/jc/blob/dev/jc/parsers/last.py

**Comment by kellyjonbrazil** [comment]
This can be tested by placing a file called `last.py` under `$HOME/.local/share/jc/jcparsers` with the code from the link above.

**Comment by zerolagtime** [comment]
support both `crash` and `down`
```diff
197c197
<                 if len(linedata) > 9 and linedata[9] != 'crash':
---
>                 if len(linedata) > 9 and linedata[9] != 'crash' and linedata[9] != 'down':
200c200
<                 if len(linedata) > 9 and linedata[9] == 'crash':
---
>                 if len(linedata) > 9 and ( linedata[9] == 'crash' or linedata[9] == 'down'):
```
Example line that broke the previous version with `last -F`:
```txt
root     pts/0        developer-pc.exa Mon Mar 23 16:44:55 2020 - down                      (00:00)
```

Other than that fix, it looks beautiful!

**Comment by kellyjonbrazil** [comment]
right-on! I'll make the change and push it to v1.14.1 in the next couple days. Thanks for testing!

**Comment by zerolagtime** [comment]
bigger patch that leverages the `-F` option and adds 
* `login_epoch` is the login time in seconds since epoch
* `logout_epoch` (if a valid date) for the logout time
* `duration_seconds` as an integer if both login_epoch and logout_epoch came up with date
* `duration` now turns "2+01:06" into "49:06"
* a translation that takes an X-Windows/GUI login and prefixes it with CONSOLE for easier matching on systems with multiple displays (like VMware virtual desktops)

These should all be easier to parse by a database or other module.
```diff
83c83
< 
---
> from datetime import datetime
135a136,138
>         if 'hostname' in entry and entry['hostname'][0] == ":": 
>             entry['hostname'] = f'CONSOLE{entry["hostname"]}'
> 
141a145,158
>         if 'login' in entry and re.match(r'.*\d\d:\d\d:\d\d \d\d\d\d.*',entry['login']):
>             entry['login_epoch'] = datetime.strptime(entry['login'], '%a %b %d %H:%M:%S %Y').strftime('%s')
> 
>         if 'logout' in entry and re.match(r'.*\d\d:\d\d:\d\d \d\d\d\d.*',entry['logout']):
>             entry['logout_epoch'] = datetime.strptime(entry['logout'], '%a %b %d %H:%M:%S %Y').strftime('%s')
> 
>         if 'login_epoch' in entry and 'logout_epoch' in entry:
>             entry['duration_seconds'] = int(entry['logout_epoch']) - int(entry['login_epoch'])
> 
>         if 'duration' in entry and re.match(r'^\d+\+', entry['duration']):
>             m = re.match(r'^(?P<days>\d+)\+(?P<hours>\d\d):(?P<minutes>\d\d)', entry['duration'])
>             days, hours, minutes = m.groups()
>             entry['duration'] = f'{int(days)*24 + int(hours)}:{minutes}'
> 
197c214
<                 if len(linedata) > 9 and linedata[9] != 'crash':
---
>                 if len(linedata) > 9 and linedata[9] != 'crash' and linedata[9] != 'down':
200c217
<                 if len(linedata) > 9 and linedata[9] == 'crash':
---
>                 if len(linedata) > 9 and ( linedata[9] == 'crash' or linedata[9] == 'down'):
```

**Comment by kellyjonbrazil** [comment]
Great stuff - I'll take a closer look and merge tomorrow. Good night!

**Comment by zerolagtime** [comment]
Thanks for writing this. I had to add a “test for __main__” block at the bottom in order to run a problem child line through my debugger.   What is your normal approach for debugging just one plugin?

**Comment by zerolagtime** [comment]
* Straight keyboard/mouse login from gnome: 
```txt
root     pts/0        :0               Thu Dec 17 11:45 - 11:45  (00:00)
```
* Login on a text console for Linux - note that column three is missing: 
```txt
user1234 tty2                          Thu Dec 17 10:37 - 10:38  (00:00)
``` 
(this works on the pre-patch plugin)
* I've also seen some durations from last like "-4:-56" which pass through, but I have no idea what it means in reality
* Here are some other interesting outputs from CentOS 7.9, :
```txt
reboot   system boot  3.10.0-1127.19.1 Tue Dec 29 08:53:03 2020 - Wed Jan  6 09:28:15 2021 (8+00:35)   
user1234 pts/0        :0               Tue Dec 29 13:50:50 2020 - crash                     (-4:-57)   
user1234 :0           :0               Tue Dec 29 13:49:58 2020 - crash                     (-4:-56)   
user1234 pts/1        :pts/0:S.0       Mon Dec 28 17:24:33 2020 - Mon Dec 28 17:24:46 2020  (00:00)    
user1234 pts/0        :0               Mon Dec 28 17:24:10 2020 - Mon Dec 28 17:25:01 2020  (00:00)    
user1234 pts/2        fe80::fd8:7309:4 Mon Dec 28 12:43:33 2020 - Mon Dec 28 13:14:22 2020  (00:30) 
```
the `:pts/0:S.0` is apparently a [screen](https://linux.die.net/man/1/screen) session and I could probably replace "CONSOLE:pts" with "SCREEN:pts" for clarity.

I also fixed a bug in my patch above that wasn't checking to see if hostname was None.

**Comment by kellyjonbrazil** [comment]
> Thanks for writing this. I had to add a “test for __main__” block at the bottom in order to run a problem child line through my debugger. What is your normal approach for debugging just one plugin?
> […](#)
> On Jan 6, 2021, at 1:23 AM, Kelly Brazil ***@***.***> wrote: ﻿ Great stuff - I'll take a closer look and merge tomorrow. Good night! — You are receiving this because you commented. Reply to this email directly, view it on GitHub, or unsubscribe.

You know, I can't remember using a debugger with `jc` - I typically use the `-dd` command argument and/or print statements since many of the parsers are pretty simple. I typically sparingly use `pdb` on my other projects, if needed, but I can't remember using it with `jc` - just a lot of trial and error and unit tests. :)

**Comment by zerolagtime** [comment]
> You know, I can't remember using a debugger with `jc` - I typically use the `-dd` command argument and/or print statements since many of the parsers are pretty simple. I typically sparingly use `pdb` on my other projects, if needed, but I can't remember using it with `jc` - just a lot of trial and error and unit tests. :)

I added the section below and then VScode was able to easily start a debugger anywhere I wanted it.
```python
if __name__ == "__main__":
    print("Testing: last (login/logout)")
    # "last" on Centos 7.9 x86_654
    data = "user    pts/4        :101             Wed Sep  9 13:05 - 13:05  (00:00)"
    raw = parse(data=data)
    if raw[0] != {'duration': '00:00', 'hostname': ':101', 'login': 'Wed Sep 9 13:05', 'logout': '13:05', 'tty': 'pts/4', 'user': 'user'}:
        raise ValueError
```
Interactive debuggers are the only way I can stay sane.  Plus I can put a breakpoint on my test case and then step through the handling of just that one thing that fails.  Variable inspection, both graphically and via REPL, are my go-to tools.

**Comment by kellyjonbrazil** [comment]
BTW, do you mind if I credit you for the enhancements in the `info` class? That info will show up with `jc -a`.

**Comment by zerolagtime** [comment]
Fine by me. Thanks!

**Comment by kellyjonbrazil** [comment]
Are you cool with me converting the following to ints?
```
"login_epoch": "1565891846",
"logout_epoch": "1565895402",
```

**Comment by kellyjonbrazil** [comment]
Strange - the epoch conversions are producing different results on Ubuntu vs. other platforms. CI/CD tests are failing:
```
..............................................................................................................................F.......F...................................................................................................................................................................................................................................................................................
======================================================================
FAIL: test_last_F_ubuntu_20_4 (test_last.MyTests)
Test 'last -F' on Ubuntu 20.4
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/jc/jc/tests/test_last.py", line 138, in test_last_F_ubuntu_20_4
    self.assertEqual(jc.parsers.last.parse(self.ubuntu_20_4_last_F, quiet=True), self.ubuntu_20_4_last_F_json)
AssertionError: Lists differ: [{'us[139 chars] 1609877031}, {'user': 'kbrazil', 'tty': 'pts/[677 chars]308}] != [{'us[139 chars] 1609905831}, {'user': 'kbrazil', 'tty': 'pts/[677 chars]108}]

First differing element 0:
{'use[87 chars] 2021', 'logout': 'still logged in', 'login_epoch': 1609877031}
{'use[87 chars] 2021', 'logout': 'still logged in', 'login_epoch': 1609905831}

Diff is 1536 characters long. Set self.maxDiff to None to see it.

======================================================================
FAIL: test_last_wF_centos_7_7 (test_last.MyTests)
Test 'last -wF' on Centos 7.7
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/jc/jc/tests/test_last.py", line 162, in test_last_wF_centos_7_7
    self.assertEqual(jc.parsers.last.parse(self.centos_7_7_last_wF, quiet=True), self.centos_7_7_last_wF_json)
AssertionError: Lists differ: [{'us[152 chars]1609856964}, {'user': 'kbrazil', 'tty': 'tty1'[39910 chars]849}] != [{'us[152 chars]1609885764}, {'user': 'kbrazil', 'tty': 'tty1'[39910 chars]849}]

First differing element 0:
{'use[99 chars] 2021', 'logout': 'still logged in', 'login_epoch': 1609856964}
{'use[99 chars] 2021', 'logout': 'still logged in', 'login_epoch': 1609885764}

Diff is 75034 characters long. Set self.maxDiff to None to see it.

----------------------------------------------------------------------
Ran 446 tests in 4.922s

FAILED (failures=2)
```
I'll investigate - not sure if there are any known issues with `datetime` on Ubuntu or maybe there is something with the environment on the Ubuntu instances in Github actions causing the date conversions to be different?

**Comment by kellyjonbrazil** [comment]
Changes merged into v1.14.1, now available on pypi.

**Comment by zerolagtime** [comment]
> Looks like exactly 28,800 seconds (8 hour difference)

My best guess is that one of them is either overriding the TZ (timezone) variable or has none set.  I have both types of boxes and they both report dates in `last` with my local timezone offset.
